src/frontend/src/heartbeat.rs

```diff
@@ -0,0 +1,125 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use api::v1::meta::{HeartbeatRequest, HeartbeatResponse, RequestHeader, Role};
+use common_telemetry::tracing::trace;
+use common_telemetry::{error, info};
+use meta_client::client::{HeartbeatSender, HeartbeatStream, MetaClient};
+use snafu::ResultExt;
+
+use crate::error;
+use crate::error::Result;
+
+#[derive(Clone)]
+pub struct HeartbeatTask {
+    meta_client: Arc<MetaClient>,
+    report_interval: u64,
+    retry_interval: u64,
+}
+
+impl HeartbeatTask {
+    pub fn new(meta_client: Arc<MetaClient>, report_interval: u64, retry_interval: u64) -> Self {
+        HeartbeatTask {
+            meta_client,
+            report_interval,
+            retry_interval,
+        }
+    }
+
+    pub async fn start(&self) -> Result<()> {
+        let (req_sender, resp_stream) = self
+            .meta_client
+            .heartbeat()
+            .await
+            .context(error::CreateMetaHeartbeatStreamSnafu)?;
+
+        info!("A heartbeat connection has been established with metasrv");
+
+        self.start_handle_resp_stream(resp_stream);
+
+        self.start_heartbeat_report(req_sender);
+
+        Ok(())
+    }
+
+    fn start_handle_resp_stream(&self, mut resp_stream: HeartbeatStream) {
+        let capture_self = self.clone();
+        let retry_interval = self.retry_interval;
+
+        common_runtime::spawn_bg(async move {
+            loop {
+                match resp_stream.message().await {
+                    Ok(Some(resp)) => capture_self.handle_response(resp).await,
+                    Ok(None) => break,
+                    Err(e) => {
+                        error!("Occur error while reading heartbeat response, err: {}", e);
+
+                        capture_self
+                            .start_with_retry(Duration::from_secs(retry_interval))
+                            .await;
+
+                        break;
+                    }
+                }
+            }
+        });
+    }
+
+    fn start_heartbeat_report(&self, req_sender: HeartbeatSender) {
+        let role = Role::Frontend as i32;
+        let report_interval = self.report_interval;
+
+        common_runtime::spawn_bg(async move {
+            let header = RequestHeader {
+                role,
+                ..Default::default()
+            };
+
+            loop {
+                let req = HeartbeatRequest {
+                    header: Some(header.clone()),
+                    ..Default::default()
+                };
+
+                if let Err(e) = req_sender.send(req).await {
+                    error!("Failed to send heartbeat to metasrv, error: {:?}", e);
+                    break;
```

Comment:
Seems you are relying `start_with_retry` invocation in another task to retry heartbeat reporting. Is it guarantee to happen if `req_sender.send` failed?

Reply:
yes